doc/contribute/coding_guidelines/index.rst

```diff
@@ -1297,6 +1297,191 @@ macros. Given that Zephyr uses a fork of the corresponding upstream for each
 module, it is always possible to patch the macro implementation in each module
 to avoid collisions.
 
+.. _coding_guideline_libc_usage_restrictions_in_zephyr_kernel:
+
+Rule A.4: C Standard Library Usage Restrictions in Zephyr Kernel
+================================================================
+
+Severity
+--------
+
+Required
+
+Description
+-----------
+
+The use of the C standard library functions and macros in the Zephyr kernel
+shall be limited to the following functions and macros from the ISO/IEC
+9899:2011 standard, also known as C11, and their extensions:
+
+.. csv-table:: List of allowed libc functions and macros in the Zephyr kernel
+   :header: Function,Source
+   :widths: auto
+
+   abort(),ISO/IEC 9899:2011
+   abs(),ISO/IEC 9899:2011
+   aligned_alloc(),ISO/IEC 9899:2011
```

Comment:
@stephanosio - iirc the items in this list are permitted (although in some cases additional justification should be made).

Just curious though - does it make sense to use std c allocation and free functions inside the kernel, or would it be better to insist that k_ variants be used for the kernel?

My thoughts are that the kernel and most native zephyr components should use k_ variants for heap allocations and frees.

It obviously makes sense to provide the std libc interface for allocation as a compatibility layer with libraries, etc.

Just thought I would mention - it's probably still covered in the sense that any use is technically justified.

Reply:
> Just curious though - does it make sense to use std c allocation and free functions inside the kernel, or would it be better to insist that k_ variants be used for the kernel?

It is always recommended to use the `k_malloc` instead of the libc `malloc` for Zephyr native applications, including the kernel itself -- that is actually documented here:

https://github.com/zephyrproject-rtos/zephyr/blob/6697c5aa0bb296ad008ef8f728df254954764962/doc/develop/languages/c/index.rst?plain=1#L112-L120

Also I have already addressed the concern raised in https://github.com/zephyrproject-rtos/zephyr/pull/57598#issuecomment-1550089122 and added the following:

```
Note that the use of the functions listed above are subject to secure and safe
coding practices and it should not be assumed that their use in the Zephyr
kernel is unconditionally permitted by being listed in this rule.
```

Reply:
I thought the `k_malloc` family could only be used in kernel mode, not userspace, and that was why there was a separation between it and `malloc`. Otherwise, the common C library would have just used `k_malloc` to implement `malloc', right?

Reply:
> I thought the `k_malloc` family could only be used in kernel mode, not userspace

Correct. The exact recommendation is to use the "memory management API," which `k_malloc` is part of.

> that was why there was a separation between it and `malloc`.

No, the `malloc` is there for supporting portable applications using the libc API. If targeting userspace, Zephyr native applications should define their own memory domains/partitions/heap, use the applicable userspace-safe APIs (e.g. `sys_heap`).

Reply:
Yes, Zephyr native code would use native memory APIs. My point was that as `malloc()` needs to support non-native code in both kernel and userspace allocations and so cannot simply use `k_malloc()`.

Reply:
> Yes, Zephyr native code would use native memory APIs. My point was that as `malloc()` needs to support non-native code in both kernel and userspace allocations and so cannot simply use `k_malloc()`.

Yes.